File: Scrapers/alphareadscraper/scraper.js
const { chromium } = require('playwright');
require('dotenv').config();

// Utility function to get Central Time
function getCentralTime() {
  return new Date().toLocaleString('en-US', {
    timeZone: 'America/Chicago',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false
  });
}

async function scrapeAlphaRead() {
  // Check if login credentials are provided
  const email = process.env.ALPHAREAD_EMAIL;
  const password = process.env.ALPHAREAD_PASSWORD;
  const enableLogin = process.env.ENABLE_LOGIN === 'true';
  const showBrowser = process.env.SHOW_BROWSER === 'true';
  
  if (enableLogin && (!email || !password || email === 'your.email@example.com')) {
    console.log('⚠️  Login enabled but credentials not properly configured in .env file');
    console.log('Please update ALPHAREAD_EMAIL and ALPHAREAD_PASSWORD in .env file');
    return;
  }
  
  // Launch browser
  const browser = await chromium.launch({ 
    headless: !showBrowser, // Show browser if SHOW_BROWSER=true in .env
    slowMo: 1000 // Slow down operations by 1 second for better observation
  });
  
  try {
    const context = await browser.newContext({
      // Set user agent to avoid bot detection
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    });
    
    const page = await context.newPage();
    
    console.log('🚀 Starting to scrape Alpha Read...');
    
    // Navigate to the Alpha Read sign-in page
    await page.goto('https://alpharead.alpha.school/signin', {
      waitUntil: 'networkidle',
      timeout: 30000
    });
    
    console.log('📄 Page loaded successfully');
    
    // Perform login if enabled and credentials are provided
    if (enableLogin && email && password) {
      console.log('🔐 Attempting to login...');
      
      try {
        // Wait for the form to be fully loaded
        await page.waitForSelector('form', { timeout: 10000 });
        console.log('📋 Login form detected');
        
        // Look for email input field with specific Alpha Read selectors
        await page.waitForSelector('input[type="email"][placeholder="you@example.com"]', { timeout: 10000 });
        
        // Fill in email using the specific selector
        const emailField = await page.$('input[type="email"][placeholder="you@example.com"]');
        if (emailField) {
          await emailField.click(); // Click to focus first
          await emailField.fill(''); // Clear any existing content
          await emailField.fill(email);
          console.log('📧 Email entered successfully');
        } else {
          throw new Error('Email field not found with expected selector');
        }
        
        // Fill in password using the specific selector
        const passwordField = await page.$('input[type="password"][placeholder="Password"]');
        if (passwordField) {
          await passwordField.click(); // Click to focus first
          await passwordField.fill(''); // Clear any existing content
          await passwordField.fill(password);
          console.log('🔑 Password entered successfully');
        } else {
          throw new Error('Password field not found with expected selector');
        }
        
        // Small delay to ensure fields are filled
        await page.waitForTimeout(1000);
        
        // Click the specific sign in button (not the Google one)
        const signInButton = await page.$('button[type="submit"]:has-text("Sign in")');
        if (signInButton) {
          await signInButton.click();
          console.log('👆 Sign in button clicked');
          
          // Wait for navigation or login completion
          try {
            await page.waitForURL(url => !url.includes('/signin'), { timeout: 15000 });
            console.log('✅ Login successful - redirected from sign-in page');
          } catch (redirectError) {
            // Check if we're still on signin page (might indicate login failure)
            const currentUrl = page.url();
            if (currentUrl.includes('/signin')) {
              console.log('⚠️  Still on sign-in page - login might have failed');
              
              // Check for error messages
              const errorMessages = await page.$$eval('[class*="error"], [class*="alert"], .text-red-500, .text-danger', 
                elements => elements.map(el => el.textContent.trim()).filter(text => text)
              );
              
              if (errorMessages.length > 0) {
                console.log('❌ Login errors found:', errorMessages);
              }
            } else {
              console.log('✅ Login appears successful');
            }
          }
          
          // Wait a moment for any dynamic content to load after login
          await page.waitForTimeout(3000);
          
          // Navigate to Guide Dashboard
          console.log('🎯 Navigating to Guide Dashboard...');
          try {
            // Look for Guide Dashboard link using exact selector
            const guideDashboardLink = await page.$('a[href="/guide"]:has-text("Guide Dashboard")');
            if (guideDashboardLink) {
              await guideDashboardLink.click();
              console.log('📊 Clicked Guide Dashboard');
              
              // Wait for page to load
              await page.waitForTimeout(3000);
              const currentUrl = page.url();
              if (currentUrl.includes('/guide')) {
                console.log('✅ Successfully navigated to Guide Dashboard');
              } else {
                console.log(`⚠️ URL after click: ${currentUrl}`);
              }
              
              // Wait for dashboard content to load
              await page.waitForTimeout(2000);
              
              // Look for student management section
              console.log('👥 Looking for Student Management...');
              
              // Try multiple possible selectors for student management
              const studentManagementSelectors = [
                'a:has-text("view students")',
                'a:has-text("student management")',
                'a:has-text("View Students")',
                'a:has-text("Student Management")',
                'button:has-text("view students")',
                'button:has-text("student management")',
                '[href*="student"]',
                '[href*="manage"]'
              ];
              
              let studentManagementElement = null;
              for (const selector of studentManagementSelectors) {
                studentManagementElement = await page.$(selector);
                if (studentManagementElement) {
                  console.log(`📋 Found student management with selector: ${selector}`);
                  break;
                }
              }
              
              if (studentManagementElement) {
                await studentManagementElement.click();
                console.log('👥 Clicked on student management section');
                
                // Wait for student management page to load
                await page.waitForTimeout(3000);
                console.log('✅ Successfully navigated to Student Management');
              } else {
                console.log('⚠️  Student management section not found with known selectors');
                
                // Let's examine what's available on the Guide Dashboard
                const availableLinks = await page.$$eval('a, button', elements => 
                  elements.map(el => ({
                    text: el.textContent.trim(),
                    href: el.href || null,
                    tag: el.tagName.toLowerCase()
                  })).filter(item => item.text)
                );
                
                console.log('🔍 Available navigation options on Guide Dashboard:');
                availableLinks.forEach(link => 
                  console.log(`  ${link.tag.toUpperCase()}: "${link.text}" ${link.href ? '-> ' + link.href : ''}`)
                );
              }
              
            } else {
              console.log('❌ Guide Dashboard link not found');
              
              // Let's see what navigation options are available
              const navLinks = await page.$$eval('a', links => 
                links.map(link => ({
                  text: link.textContent.trim(),
                  href: link.href
                })).filter(link => link.text)
              );
              
              console.log('🔍 Available navigation links:');
              navLinks.forEach(link => console.log(`  "${link.text}" -> ${link.href}`));
            }
            
          } catch (navError) {
            console.log('❌ Navigation error:', navError.message);
            console.log('📄 Continuing with current page scraping...');
          }
          
        }
      } catch (loginError) {
        console.log('❌ Login failed:', loginError.message);
        console.log('📄 Continuing with page scraping...');
      }
    }
    
    // Extract page data
    const pageData = await page.evaluate(() => {
      const data = {};
      
      // Get page title
      data.title = document.title;
      
      // Get page URL
      data.url = window.location.href;
      
      // Look for the main heading
      const mainHeading = document.querySelector('h1');
      data.mainHeading = mainHeading ? mainHeading.textContent.trim() : null;
      
      // Get all headings
      const headings = Array.from(document.querySelectorAll('h1, h2, h3, h4, h5, h6'))
        .map(h => ({
          tag: h.tagName.toLowerCase(),
          text: h.textContent.trim()
        }));
      data.headings = headings;
      
      // Look for form elements
      const forms = Array.from(document.querySelectorAll('form')).map(form => {
        const inputs = Array.from(form.querySelectorAll('input')).map(input => ({
          type: input.type,
          name: input.name,
          placeholder: input.placeholder,
          id: input.id,
          required: input.required
        }));
        
        const buttons = Array.from(form.querySelectorAll('button')).map(btn => ({
          text: btn.textContent.trim(),
          type: btn.type,
          id: btn.id,
          className: btn.className
        }));
        
        return { inputs, buttons };
      });
      data.forms = forms;
      
      // Look for sign-in options
      const signInButtons = Array.from(document.querySelectorAll('button, a'))
        .filter(el => el.textContent.toLowerCase().includes('sign') || 
                     el.textContent.toLowerCase().includes('google') ||
                     el.textContent.toLowerCase().includes('login'))
        .map(el => ({
          text: el.textContent.trim(),
          tag: el.tagName.toLowerCase(),
          href: el.href || null,
          className: el.className
        }));
      data.signInOptions = signInButtons;
      
      // Get all links
      const links = Array.from(document.querySelectorAll('a'))
        .map(link => ({
          text: link.textContent.trim(),
          href: link.href,
          target: link.target
        }))
        .filter(link => link.text && link.href);
      data.links = links;
      
      // Get meta information
      const metaTags = Array.from(document.querySelectorAll('meta'))
        .map(meta => ({
          name: meta.name,
          property: meta.getAttribute('property'),
          content: meta.content
        }))
        .filter(meta => meta.name || meta.property);
      data.metaTags = metaTags;
      
      return data;
    });
    
    // Take a screenshot
    await page.screenshot({ 
      path: 'alpharead-screenshot.png',
      fullPage: true 
    });
    
    console.log('📸 Screenshot saved as alpharead-screenshot.png');
    
    // Check if there are any input fields (for potential form interaction)
    const inputFields = await page.$$eval('input', inputs => 
      inputs.map(input => ({
        type: input.type,
        name: input.name,
        placeholder: input.placeholder,
        required: input.required
      }))
    );
    
    // Log the scraped data
    console.log('\n📊 SCRAPED DATA:');
    console.log('='.repeat(50));
    console.log(`Title: ${pageData.title}`);
    console.log(`URL: ${pageData.url}`);
    console.log(`Main Heading: ${pageData.mainHeading}`);
    
    if (pageData.headings.length > 0) {
      console.log('\n📝 Headings found:');
      pageData.headings.forEach(h => console.log(`  ${h.tag.toUpperCase()}: ${h.text}`));
    }
    
    if (pageData.signInOptions.length > 0) {
      console.log('\n🔐 Sign-in options found:');
      pageData.signInOptions.forEach(option => console.log(`  ${option.tag.toUpperCase()}: ${option.text}`));
    }
    
    if (inputFields.length > 0) {
      console.log('\n📝 Input fields found:');
      inputFields.forEach(field => console.log(`  ${field.type}: ${field.placeholder || field.name || 'No placeholder'}`));
    }
    
    if (pageData.links.length > 0) {
      console.log('\n🔗 Links found:');
      pageData.links.forEach(link => console.log(`  "${link.text}" -> ${link.href}`));
    }
    
    // Save data to JSON file
    const fs = require('fs');
    
    // Add current URL and timestamp to the data
    pageData.currentUrl = page.url();
    pageData.scrapedAt = getCentralTime();
    pageData.navigationPath = enableLogin ? 'Login -> Guide Dashboard -> Student Management' : 'Direct page access';
    
    fs.writeFileSync('alpharead-data.json', JSON.stringify(pageData, null, 2));
    console.log('\n💾 Data saved to alpharead-data.json');
    console.log(`📍 Final URL: ${pageData.currentUrl}`);
    
    return pageData;
    
  } catch (error) {
    console.error('❌ Error occurred:', error.message);
    throw error;
  } finally {
    await browser.close();
    console.log('🏁 Browser closed');
  }
}

// Run the scraper
if (require.main === module) {
  scrapeAlphaRead()
    .then(data => {
      console.log('\n✅ Scraping completed successfully!');
      console.log(`📊 Total data points collected: ${Object.keys(data).length}`);
    })
    .catch(error => {
      console.error('💥 Scraping failed:', error);
      process.exit(1);
    });
}

module.exports = { scrapeAlphaRead }; 

File: Scrapers/fmpscraper/scraper.py
#!/usr/bin/env python3
"""
FastMath Pro Downloads Scraper
A Selenium-based web scraper for app.fastmath.pro/admin/downloads
"""

import time
import logging
import os
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.common.exceptions import TimeoutException, NoSuchElementException
from webdriver_manager.chrome import ChromeDriverManager
import json
import csv
from datetime import datetime
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('scraper.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)


class FastMathProScraper:
    def __init__(self, headless=True, timeout=30):
        """
        Initialize the scraper with Chrome WebDriver
        
        Args:
            headless (bool): Run browser in headless mode
            timeout (int): Default timeout for web elements
        """
        self.timeout = timeout
        self.driver = None
        self.setup_driver(headless)
        
    def setup_driver(self, headless):
        """Setup Chrome WebDriver with appropriate options"""
        try:
            chrome_options = Options()
            if headless:
                chrome_options.add_argument("--headless")
            chrome_options.add_argument("--no-sandbox")
            chrome_options.add_argument("--disable-dev-shm-usage")
            chrome_options.add_argument("--disable-gpu")
            chrome_options.add_argument("--window-size=1920,1080")
            chrome_options.add_argument("--user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")
            
            # Use ChromeDriverManager to automatically download and manage ChromeDriver
            service = Service(ChromeDriverManager().install())
            self.driver = webdriver.Chrome(service=service, options=chrome_options)
            self.driver.implicitly_wait(self.timeout)
            logger.info("Chrome WebDriver initialized successfully")
            
        except Exception as e:
            logger.error(f"Failed to initialize WebDriver: {str(e)}")
            raise
    
    def login(self, username=None, password=None):
        """
        Attempt to login to the admin panel
        
        Args:
            username (str): Username (if None, reads from environment)
            password (str): Password (if None, reads from environment)
            
        Returns:
            bool: True if login successful, False otherwise
        """
        try:
            # Get credentials from environment if not provided
            if username is None:
                username = os.getenv('USERNAME')
            if password is None:
                password = os.getenv('PASSWORD')
            
            if not username or not password:
                logger.error("No credentials provided. Please set USERNAME and PASSWORD in .env file")
                logger.error(f"USERNAME env var: {'SET' if os.getenv('USERNAME') else 'NOT SET'}")
                logger.error(f"PASSWORD env var: {'SET' if os.getenv('PASSWORD') else 'NOT SET'}")
                return False
            
            logger.info("Attempting to login...")
            
            # Look for common login form elements
            login_selectors = [
                "input[type='email']",
                "input[name='email']",
                "input[name='username']", 
                "input[id='email']",
                "input[id='username']"
            ]
            
            password_selectors = [
                "input[type='password']",
                "input[name='password']",
                "input[id='password']"
            ]
            
            submit_selectors = [
                "button[type='submit']",
                "input[type='submit']",
                "button:contains('Login')",
                "button:contains('Sign in')",
                ".login-button",
                ".submit-button"
            ]
            
            # Find username field
            username_field = None
            for selector in login_selectors:
                try:
                    username_field = self.driver.find_element(By.CSS_SELECTOR, selector)
                    break
                except NoSuchElementException:
                    continue
            
            if not username_field:
                logger.error("Could not find username/email field")
                return False
            
            # Find password field
            password_field = None
            for selector in password_selectors:
                try:
                    password_field = self.driver.find_element(By.CSS_SELECTOR, selector)
                    break
                except NoSuchElementException:
                    continue
            
            if not password_field:
                logger.error("Could not find password field")
                return False
            
            # Fill in credentials
            username_field.clear()
            username_field.send_keys(username)
            
            password_field.clear()
            password_field.send_keys(password)
            
            # Find and click submit button
            submit_button = None
            for selector in submit_selectors:
                try:
                    submit_button = self.driver.find_element(By.CSS_SELECTOR, selector)
                    break
                except NoSuchElementException:
                    continue
            
            if submit_button:
                submit_button.click()
            else:
                # Try submitting the form by pressing Enter
                password_field.send_keys("\n")
            
            # Wait for login to process
            time.sleep(3)
            
            # Check if login was successful by looking for login indicators
            if self.is_login_required():
                logger.error("Login appears to have failed - still seeing login page")
                return False
            else:
                logger.info("Login appears successful")
                return True
                
        except Exception as e:
            logger.error(f"Error during login: {str(e)}")
            return False

    def scrape_downloads_page(self, url="https://app.alphamath.school/admin/downloads"):
        """
        Scrape the FastMath Pro admin downloads page
        
        Args:
            url (str): The URL to scrape
            
        Returns:
            dict: Scraped data
        """
        scraped_data = {
            "url": url,
            "timestamp": datetime.now().isoformat(),
            "status": "failed",
            "data": [],
            "error": None
        }
        
        try:
            logger.info(f"Navigating to: {url}")
            self.driver.get(url)
            
            # Wait for page to load
            WebDriverWait(self.driver, self.timeout).until(
                EC.presence_of_element_located((By.TAG_NAME, "body"))
            )
            
            # Check if login is required and attempt login
            if self.is_login_required():
                logger.warning("Login page detected. Attempting authentication...")
                if self.login():
                    logger.info("Login successful, proceeding with scraping...")
                    # Navigate to the downloads page again after login
                    self.driver.get(url)
                    WebDriverWait(self.driver, self.timeout).until(
                        EC.presence_of_element_located((By.TAG_NAME, "body"))
                    )
                    
                    # Check if still on login page
                    if self.is_login_required():
                        scraped_data["error"] = "Authentication failed - still on login page"
                        return scraped_data
                else:
                    scraped_data["error"] = "Authentication failed - could not login"
                    return scraped_data
            
            # Get page title
            page_title = self.driver.title
            logger.info(f"Page title: {page_title}")
            
            # Look for and click "Back to Admin Dashboard" button to navigate to student dashboard
            logger.info("Looking for 'Back to Admin Dashboard' button...")
            try:
                back_button = self.driver.find_element(By.XPATH, "//a[contains(text(), 'Back to Admin Dashboard')]")
                logger.info(f"Found back button: {back_button.text}")
                logger.info("Clicking 'Back to Admin Dashboard' button...")
                back_button.click()
                time.sleep(5)  # Wait for navigation
                
                # Wait for dashboard to load
                WebDriverWait(self.driver, self.timeout).until(
                    EC.presence_of_element_located((By.TAG_NAME, "table"))
                )
                logger.info("Successfully navigated to admin dashboard")
                
                # Now scrape dashboard student data
                logger.info("Scraping dashboard student data...")
                student_data = self.extract_dashboard_students()
                
            except Exception as e:
                logger.error(f"Could not navigate to dashboard: {str(e)}")
                student_data = []
            
            # Scrape download links and information (if still on downloads page)
            downloads = self.extract_download_data()
            
            # Scrape general page information
            page_info = self.extract_page_info()
            
            scraped_data.update({
                "status": "success",
                "page_title": page_title,
                "page_info": page_info,
                "downloads": downloads,
                "total_downloads": len(downloads),
                "students": student_data if 'student_data' in locals() else [],
                "total_students": len(student_data) if 'student_data' in locals() else 0
            })
            
            logger.info(f"Successfully scraped {len(downloads)} download items")
            
        except TimeoutException:
            error_msg = f"Timeout waiting for page to load: {url}"
            logger.error(error_msg)
            scraped_data["error"] = error_msg
            
        except Exception as e:
            error_msg = f"Error scraping page: {str(e)}"
            logger.error(error_msg)
            scraped_data["error"] = error_msg
            
        return scraped_data
    
    def is_login_required(self):
        """Check if the page requires login"""
        login_indicators = [
            "login", "sign in", "username", "password", 
            "authentication", "access denied", "unauthorized"
        ]
        
        page_text = self.driver.page_source.lower()
        return any(indicator in page_text for indicator in login_indicators)
    
    def extract_download_data(self):
        """Extract download links and related information"""
        downloads = []
        
        try:
            # Common selectors for download links
            download_selectors = [
                "a[href*='download']",
                "a[href*='.pdf']",
                "a[href*='.zip']",
                "a[href*='.exe']",
                "a[href*='.dmg']",
                ".download-link",
                ".download-item",
                "[data-download]"
            ]
            
            for selector in download_selectors:
                try:
                    elements = self.driver.find_elements(By.CSS_SELECTOR, selector)
                    for element in elements:
                        download_info = {
                            "text": element.text.strip(),
                            "href": element.get_attribute("href"),
                            "title": element.get_attribute("title"),
                            "data_attrs": self.get_data_attributes(element)
                        }
                        if download_info["href"] and download_info not in downloads:
                            downloads.append(download_info)
                except NoSuchElementException:
                    continue
            
            # Also look for buttons or divs that might contain download information
            button_selectors = [
                "button[class*='download']",
                ".btn-download",
                "[role='button'][aria-label*='download']"
            ]
            
            for selector in button_selectors:
                try:
                    elements = self.driver.find_elements(By.CSS_SELECTOR, selector)
                    for element in elements:
                        download_info = {
                            "text": element.text.strip(),
                            "type": "button",
                            "onclick": element.get_attribute("onclick"),
                            "data_attrs": self.get_data_attributes(element)
                        }
                        downloads.append(download_info)
                except NoSuchElementException:
                    continue
                    
        except Exception as e:
            logger.error(f"Error extracting download data: {str(e)}")
        
        return downloads
    
    def extract_page_info(self):
        """Extract general page information"""
        page_info = {}
        
        try:
            # Get all headings
            headings = []
            for level in range(1, 7):
                h_elements = self.driver.find_elements(By.TAG_NAME, f"h{level}")
                for h in h_elements:
                    if h.text.strip():
                        headings.append({
                            "level": level,
                            "text": h.text.strip()
                        })
            
            page_info["headings"] = headings
            
            # Get all tables if any
            tables = []
            table_elements = self.driver.find_elements(By.TAG_NAME, "table")
            for i, table in enumerate(table_elements):
                table_data = {
                    "index": i,
                    "rows": len(table.find_elements(By.TAG_NAME, "tr")),
                    "html": table.get_attribute("outerHTML")[:500] + "..." if len(table.get_attribute("outerHTML")) > 500 else table.get_attribute("outerHTML")
                }
                tables.append(table_data)
            
            page_info["tables"] = tables
            
            # Get meta information
            page_info["url"] = self.driver.current_url
            page_info["title"] = self.driver.title
            
        except Exception as e:
            logger.error(f"Error extracting page info: {str(e)}")
        
        return page_info
    
    def get_data_attributes(self, element):
        """Extract data-* attributes from an element"""
        data_attrs = {}
        try:
            # Get all attributes
            attrs = self.driver.execute_script(
                "var items = {}; for (index = 0; index < arguments[0].attributes.length; ++index) { items[arguments[0].attributes[index].name] = arguments[0].attributes[index].value }; return items;",
                element
            )
            
            # Filter data attributes
            data_attrs = {k: v for k, v in attrs.items() if k.startswith('data-')}
        except Exception as e:
            logger.error(f"Error getting data attributes: {str(e)}")
        
        return data_attrs
    
    def extract_dashboard_students(self):
        """Extract student information from the admin dashboard"""
        students = []
        
        try:
            logger.info("Extracting student data from dashboard...")
            
            # Scroll to load all students
            logger.info("Scrolling to load all students...")
            last_height = self.driver.execute_script("return document.body.scrollHeight")
            scroll_count = 0
            
            while scroll_count < 20:  # Limit scrolls to prevent infinite loop
                self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
                time.sleep(2)
                new_height = self.driver.execute_script("return document.body.scrollHeight")
                if new_height == last_height:
                    break
                last_height = new_height
                scroll_count += 1
            
            logger.info(f"Completed {scroll_count} scrolls to load content")
            
            # Find the table with student data
            table = self.driver.find_element(By.TAG_NAME, "table")
            rows = table.find_elements(By.TAG_NAME, "tr")
            
            logger.info(f"Found table with {len(rows)} rows")
            
            for i, row in enumerate(rows):
                try:
                    cells = row.find_elements(By.TAG_NAME, "td")
                    if len(cells) >= 2:  # Make sure we have at least name and some data
                        student_name = cells[0].text.strip()
                        
                        # Skip empty names or header rows
                        if student_name and not student_name.isdigit() and len(student_name) > 2:
                            student_info = {
                                "index": i,
                                "name": student_name,
                                "row_data": [cell.text.strip() for cell in cells]
                            }
                            students.append(student_info)
                            
                except Exception as e:
                    logger.debug(f"Error processing row {i}: {str(e)}")
                    continue
            
            logger.info(f"Successfully extracted {len(students)} students from dashboard")
            
        except Exception as e:
            logger.error(f"Error extracting dashboard students: {str(e)}")
        
        return students
    
    def save_results(self, data, format='json'):
        """Save scraped results to file"""
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        
        if format == 'json':
            filename = f"fastmath_downloads_{timestamp}.json"
            with open(filename, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
            logger.info(f"Results saved to {filename}")
            
        elif format == 'csv' and data.get('downloads'):
            filename = f"fastmath_downloads_{timestamp}.csv"
            with open(filename, 'w', newline='', encoding='utf-8') as f:
                if data['downloads']:
                    writer = csv.DictWriter(f, fieldnames=data['downloads'][0].keys())
                    writer.writeheader()
                    writer.writerows(data['downloads'])
            logger.info(f"Downloads saved to {filename}")
    
    def close(self):
        """Close the WebDriver"""
        if self.driver:
            self.driver.quit()
            logger.info("WebDriver closed")


def main():
    """Main function to run the scraper"""
    scraper = None
    
    try:
        # Get configuration from environment
        headless_mode = os.getenv('HEADLESS_MODE', 'true').lower() == 'true'
        timeout_seconds = int(os.getenv('TIMEOUT_SECONDS', '30'))
        
        # Initialize scraper
        logger.info("Starting FastMath Pro Downloads Scraper")
        logger.info(f"Configuration: Headless={headless_mode}, Timeout={timeout_seconds}s")
        
        # Debug environment variables
        logger.info(f"Environment check - USERNAME: {'SET' if os.getenv('USERNAME') else 'NOT SET'}")
        logger.info(f"Environment check - PASSWORD: {'SET' if os.getenv('PASSWORD') else 'NOT SET'}")
        logger.info(f"Environment check - SUPABASE_URL: {'SET' if os.getenv('SUPABASE_URL') else 'NOT SET'}")
        logger.info(f"Environment check - CI: {os.getenv('CI', 'NOT SET')}")
        scraper = FastMathProScraper(headless=headless_mode, timeout=timeout_seconds)
        
        # Scrape the downloads page
        results = scraper.scrape_downloads_page()
        
        # Print results
        print(f"\nScraping Results:")
        print(f"Status: {results['status']}")
        print(f"URL: {results['url']}")
        print(f"Timestamp: {results['timestamp']}")
        
        if results['status'] == 'success':
            print(f"Page Title: {results.get('page_title', 'N/A')}")
            print(f"Total Downloads Found: {results.get('total_downloads', 0)}")
            
            if results.get('downloads'):
                print("\nDownload Items Found:")
                for i, download in enumerate(results['downloads'][:5], 1):  # Show first 5
                    print(f"{i}. {download.get('text', 'N/A')} - {download.get('href', 'N/A')}")
                
                if len(results['downloads']) > 5:
                    print(f"... and {len(results['downloads']) - 5} more items")
        else:
            print(f"Error: {results.get('error', 'Unknown error')}")
        
        # Save results
        scraper.save_results(results, 'json')
        if results.get('downloads'):
            scraper.save_results(results, 'csv')
            
    except KeyboardInterrupt:
        logger.info("Scraping interrupted by user")
    except Exception as e:
        logger.error(f"Unexpected error: {str(e)}")
    finally:
        if scraper:
            scraper.close()


if __name__ == "__main__":
    main()